storage/src/main/java/tech/pegasys/teku/storage/server/VersionedDatabaseFactory.java

```diff
@@ -164,13 +208,61 @@ private Database createV4Database() {
    */
   private Database createV5Database() {
     try {
-      final DatabaseMetadata metaData =
-          DatabaseMetadata.init(getMetadataFile(), DatabaseMetadata.v5Defaults());
+      final V5DatabaseMetadata metaData =
+          V5DatabaseMetadata.init(getMetadataFile(), V5DatabaseMetadata.v5Defaults());
       DatabaseNetwork.init(getNetworkFile(), Constants.GENESIS_FORK_VERSION, eth1Address);
       return RocksDbDatabase.createV4(
           metricsSystem,
           metaData.getHotDbConfiguration().withDatabaseDir(dbDirectory.toPath()),
-          metaData.getArchiveDbConfiguration().withDatabaseDir(archiveDirectory.toPath()),
+          metaData.getArchiveDbConfiguration().withDatabaseDir(v5ArchiveDirectory.toPath()),
+          stateStorageMode,
+          stateStorageFrequency);
+    } catch (final IOException e) {
+      throw new DatabaseStorageException("Failed to read metadata", e);
+    }
+  }
+
+  private Database createV6Database() {
+    try {
+      final V6DatabaseMetadata defaultMetaData;
+      if (v6ArchiveDirectory.isPresent()) {
+        defaultMetaData = V6DatabaseMetadata.separateDBDefault();
+      } else {
+        defaultMetaData = V6DatabaseMetadata.singleDBDefault();
+      }
+
+      final V6DatabaseMetadata metaData =
+          V6DatabaseMetadata.init(getMetadataFile(), defaultMetaData);
+      if (defaultMetaData.isSingleDB() != metaData.isSingleDB()) {
+        throw new DatabaseStorageException(
+            "The database was originally created as "
+                + (metaData.isSingleDB() ? "Single" : "Separate")
+                + " but now accessed as "
+                + (defaultMetaData.isSingleDB() ? "Single" : "Separate"));
+      }
+
+      DatabaseNetwork.init(getNetworkFile(), Constants.GENESIS_FORK_VERSION, eth1Address);
+
+      final RocksDbConfiguration hotOrSingleDBConfiguration =
+          metaData.isSingleDB()
+              ? metaData.getSingleDbConfiguration().get().getConfiguration()
+              : metaData.getSeparateDbConfiguration().get().getHotDbConfiguration();
+
+      final Optional<RocksDbConfiguration> finalizedConfiguration =
+          v6ArchiveDirectory.map(
+              dir ->
+                  metaData
+                      .getSeparateDbConfiguration()
+                      .get()
+                      .getArchiveDbConfiguration()
+                      .withDatabaseDir(dir.toPath()));
```

Comment:
Any existing metadata should override preferences wrt single / multiple dbs:
```suggestion
          v6ArchiveDirectory
          .filter(__ -> !metaData.isSingleDB())
          .map(dir ->
                  metaData
                      .getSeparateDbConfiguration()
                      .get()
                      .getArchiveDbConfiguration()
                      .withDatabaseDir(dir.toPath()));
```

Reply:
We are throwing exception if there is a mismatch between existing metadata and presence/absence of `v6ArchiveDirectory`. So the suggested `filter()` condition would always be true. 
https://github.com/PegaSysEng/teku/blob/d004e277b9949cf59e3a9e9394767d0566168477/storage/src/main/java/tech/pegasys/teku/storage/server/VersionedDatabaseFactory.java#L227-L242

Reply:
ah - i missed that, thanks!